ethexe/cli/src/tests.rs

```diff
@@ -0,0 +1,328 @@
+
+        let router_query = RouterQuery::new(&rpc, router_address).await?;
+        let genesis_block_hash = router_query.genesis_block_hash().await?;
+
+        let query = Query::new(
+            Box::new(db.clone()),
+            &rpc,
+            router_address,
+            genesis_block_hash,
+            blob_reader.clone(),
+            10000,
+        )
+        .await?;
+
+        let processor = Processor::new(db.clone())?;
+
+        let sequencer = Sequencer::new(
+            &ethexe_sequencer::Config {
+                ethereum_rpc: rpc.clone(),
+                sign_tx_public: sequencer_public_key,
+                router_address,
+            },
+            signer.clone(),
+        )
+        .await?;
+
+        let validator = Validator::new(
+            &ethexe_validator::Config {
+                pub_key: validator_public_key,
+                router_address,
+            },
+            signer.clone(),
+        );
+
+        let observer = Observer::new(&rpc, router_address, blob_reader.clone())
+            .await
+            .expect("failed to create observer");
+
+        let rpc = ethexe_rpc::RpcService::new(9090, db.clone());
+
+        let service = Service::new_from_parts(
+            db.clone(),
+            network,
+            observer.clone(),
+            query,
+            processor,
+            signer,
+            Some(sequencer),
+            Some(validator),
+            None,
+            rpc,
+        );
+
+        let env = TestEnv {
+            db,
+            blob_reader,
+            observer,
+            ethereum,
+            router_query,
+            service: Some(service),
+        };
+
+        Ok(env)
+    }
+
+    pub fn new_listener(&self) -> Listener {
+        Listener::new(self.observer.clone())
+    }
+
+    pub async fn upload_code(&self, code: &[u8]) -> Result<(H256, CodeId)> {
+        let (tx_hash, code_id) = self
+            .ethereum
+            .router()
+            .upload_code_with_sidecar(code)
+            .await?;
+        self.blob_reader
+            .add_blob_transaction(tx_hash, code.to_vec())
+            .await;
+        Ok((tx_hash, code_id))
+    }
+}
+
+#[tokio::test(flavor = "multi_thread")]
+#[ntest::timeout(60_000)]
+async fn ping() {
+    gear_utils::init_default_logger();
+
+    let anvil = Anvil::new().try_spawn().unwrap();
+
+    let mut env = TestEnv::new(anvil.ws_endpoint()).await.unwrap();
+    let mut listener = env.new_listener();
+
+    let service = env.service.take().unwrap();
+    let service_handle = tokio::task::spawn(service.run());
+
+    let (_, code_id) = env.upload_code(demo_ping::WASM_BINARY).await.unwrap();
+
+    log::info!("📗 Waiting for code loaded");
+    listener
+        .apply_until(|event| {
+            if let Event::CodeLoaded(loaded) = event {
+                assert_eq!(loaded.code_id, code_id);
+                assert_eq!(loaded.code.as_slice(), demo_ping::WASM_BINARY);
+                Ok(Some(()))
```

Comment:
`Result<Option<R>>` is more flexible, but here `Option<R>` is enough. why can't we just have something like `events.find(predicate)` and then do assertion when event is found?

Reply:
Because we have `Observer::Event` and `BlockEvent`. One observer event contains maybe several `BlockEvent`, so we need to check them in one `call` like `apply_until_block_event`, or return all of that we need in one `call`.

Anyway, we can make find also. But let's to this in future refactorings